FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1892 @@
+  Ref LoadStackValueAtOffset_Slow(uint8_t Offset);
+  void StoreStackValueAtOffset_Slow(uint8_t Offset, Ref Value, bool SetValid = true);
+  // Update Top value in slow path for a pop
+  void UpdateTop4Pop_Slow();
+  void UpdateTop4Push_Slow();
+  // Synchronizes the current simulated stack with the actual values.
+  // Returns a new value for Top, that's synchronized between the simulated stack
+  // and the actual FPU stack.
+  Ref SynchronizeStackValues();
+  // Moves us from the fast to the slow path if ShouldMigrate is true.
+  void MigrateToSlowPathIf(bool ShouldMigrate);
+  // Top Cache Management
+  Ref GetTopWithCache_Slow();
+  Ref GetOffsetTopWithCache_Slow(uint8_t Offset);
+  void SetTopWithCache_Slow(Ref Value);
+  Ref GetX87ValidTag_Slow(uint8_t Offset);
+  // Resets fields to initial values
+  void Reset(bool AlsoSlowPath = true);
+
+  struct StackMemberInfo {
+    IR::OpSize SourceDataSize; // Size of SourceDataNode
+    IR::OpSize StackDataSize;  // Size of the data in the stack
+    IR::Ref SourceDataNode;    // Reference to the source location of the data.
+                               // In the case of a load, this is the source node of the load.
+                               // If it's not a load, then it's nullptr.
+    IR::Ref StackDataNode;     // Reference to the data in the Stack.
+    bool InterpretAsFloat {};  // True if this is a floating point value, false if integer
+  };
+
+  // StackData, TopCache need to be always properly set to ensure
+  // it reflects the current state of the FPU. This sync only makes sense while
+  // taking the fast path. Once in the slow path, these don't make sense anymore
+  // and we are syncing everything.
+
+  // Index on vector is offset to top value at start of block
+  // If slow path is true, then StackData is always empty.
+  FixedSizeStack<StackMemberInfo, 8> StackData;
+  using StackSlot = FixedSizeStack<StackMemberInfo, 8>::Slot;
+
+  void InvalidateCaches();
+  void InvalidateTopOffsetCache();
+
+  // Cache for Constants
+  // ConstantPoll[i] has IREmit->_Constant(i);
+  constexpr static size_t ConstantPoolSize = 8;
+  std::array<Ref, ConstantPoolSize> ConstantPool;
+  Ref GetConstant(ssize_t Offset);
+
+  // Cached value for Top
+  // If slowpath is false, then TopCache is nullptr.
+  std::array<Ref, 8> TopOffsetCache;
+  // Are we on the slow path?
+  // Once we enter the slow path, we never come out.
+  // This just simplifies the code atm. If there's a need to return to the fast path in the future
+  // we can implement that but I would expect that there would be very few cases where that's necessary.
+  // On the slow path TopCache is always the last obtained version of top.
+  // TopOffset is ignored
+  bool SlowPath = false;
+  // Keeping IREmitter not to pass arguments around
+  IREmitter* IREmit = nullptr;
+};
+
+inline void X87StackOptimization::InvalidateCaches() {
+  LogMan::Msg::DFmt("Invalidating caches");
```

Comment:
I expect most of the debug prints should be removed, but I don't have a good sense of this pass yet so not sure.

Reply:
Yes, that's true - they will be removed. But I will leave those until before the merge. :)

Just please ignore them until we are close to merge.